### ISSUE TITLE
Create Guidelines for MPL as class project

### ISSUE THREAD

**Issue by story645** [matplotlib/matplotlib#16702]
Not quite sure where this should go in the docs, but semi-regularly a class gets assigned to contribute to open source projects (awesome!) and then everyone gets a little frustrated (not so awesome), so I think we should add some guidelines to the docs (maybe in the contributing guide somewhere?). Something like

### Guidelines
Thank you for encouraging your students to contribute to Matplotlib. To make this a productive experience for everyone, we ask that you please:
1. Inform us at least <a week?> in advance, via either [gitter](https://gitter.im/matplotlib/matplotlib), [discourse](https://discourse.matplotlib.org/c/development/contribute/8), or the [matplotlib-dev](https://mail.python.org/mailman/listinfo/matplotlib-devel) mailing list. (I don't want this to just be a @tacaswell gets an email thing)
2. Email a class roster of github usernames to matplotlib@numfocus.org (or some other method of confidentially sharing the information)
3. Help us curate a set list of issues that you think are appropriate for your students. With enough notice, we may even be able to create a project specifically for your class. 
5. Verify that your students have read our [contributing guide](https://matplotlib.org/devel/index.html). They can ask questions on [discourse](https://discourse.matplotlib.org/c/development/contribute/8)
4. If you have more than 50? students, please provide the github handle of at least one person we can assign to review the student contributions.

**Comment by jklymak** [comment]
These tend to get started mid semester, and hence the students have something like 6 weeks to complete.  That timeline seems prohibitively short to me unless the change is quite trivial.

**Comment by anntzer** [comment]
copy-pasting my gitter post, especially in reply to (5):

> I don't think we should even claim to support this as a "special mode of contrib", even with ground rules...
> someone comes, someone writes a pr, great, it may get reviewed fast, it may get reviewed slow, we can't guarantee it to get reviewed fast enough for you to pass your class
> (and we're not going to change reviewing rules for the convenience of your class)

**Comment by tacaswell** [comment]
This has been a recurring instance for ~5 years (and there is a paper about it ithttps://scholarspace.manoa.hawaii.edu/handle/10125/60211 ) , I am worried we are over correcting.

I am :-1: on asking classes to send us rosters or curating anything specifically for the class as that is putting way more burden on us than a small burst of new contributors.

I also agree with @anntzer that we don't want to commit to doing anything special for the class (in fact not being treated special is part of the benefit to the students).   However it is useful to know when it is happening so that if we get a new contributor who is acting badly, we may be able to bounce that problem to the instructor.

**Comment by story645** [comment]
So that's why I think the roster is a good idea, precisely cause we'd be able to identify "hey it's a student" and cause this is legit not much extra work on the teachers part since they have that information.

The only reason I'm thinking projects is cause we already do this for sprints and the like (I'd do that if I was running this in one of my classes), and @timhoffm mentioned wanting to better curate issues. But even just curating issues - like anything the teacher flags as good for their students that the students don't do goes into our pool of good first issues & I really don't think it's too much to ask a teacher to contribute a little bit of upfront work to the project.

Granted I also think part of having guidelines is to spell out "we are unable to gaurantee that your students PRs will be reviewed/merged in the time frame allocated to this project"

**Comment by tacaswell** [comment]
attn @atafliovich

**Comment by story645** [comment]
So alternative to adding it to docs is making a pinned thread version of the guidelines on the discourse in the contributing section.
And @atafliovich how would you feel about somehow incentivising one or two students or someone supervisong to write a blog post about their experience for the Matplotlib blog?

**Comment by atafliovich** [comment]
I ran this course (Engineering Large Software Systems at University of Toronto Scarborough) between 2013 and 2017, and again in 2019 and now in 2020. Our collaboration has been, in my opinion, very beneficial for the students. 
 
The number of students in this course has quadrupled over the years, so this term, to make sure not too many teams try to contribute to matplotlib, I gave the students a choice between 5 or so projects. I believe 10 or fewer teams chose matplotlib.

I hadn't realised that "everyone gets a little frustrated" @story645 . I would like to apologise, both on my behalf and on behalf of my students.

@jklymak is correct: the part of the course project when the students are making contributions is 6-7 weeks long. Our academic term is 12 weeks long, so I don't think the timeline can change much. I understand that in most cases this implies that contributions are limited in scope. Although we did have a couple of notable contributions over the years.

I welcome all suggestions from you, ranging from "please, stop using our project in your course" to "please, contact us ahead of time and we will select a number of issues for your students to tackle".

I have a learning objective in mind when I ask students to select the issue(s) they want to work on. I believe they learn from the process of exploring the project. But if this does not work well for you, I can certainly change that.

There is already a requirement for the students to mention the course when communicating here. I can do more to enforce it. I can also send you a list of guthub logins of the students in the course (once I have their consent).

Whichever way we go from here, I would like to emphasise that there is absolutely no expectation on my part that you will treat contributions from my students any differently than you would any other.

**Comment by atafliovich** [comment]
@story645 Yes, of course. It would be a good experience for the student.

**Comment by story645** [comment]
Thank you for giving some I think much needed context. I think maybe the aim is somewhere between they're 💯 like any other contributor and they're babysat 'cause I hear you and others on the pedogogical value of exploring the project. The students have definitely made some valuable and much appreciated contributions over the years.

I think going forward a heads up w/ an estimated count of how many students are going to participate (if possible) & the assignment expectations would be useful as that relieves us of some pressure of feeling like grades depend on us.
 
I don't think a TA reviewing beforehand is necessary, but it would be helpful if they monitered the communications/issues/PRs and interceded on interpersonal situations outside of scope. For example can't remember when but some students started fighting over who should get credit for the PR and tried to snipe them. Situations like that at the intersection between project and assignment.

And awesome, the blog lives at https://github.com/matplotlib/matplotblog

**Comment by timhoffm** [comment]
Thanks for the background information. This helps very much in judging the situation.

As I've probably started the topic with my comments and as I'm one of the review-intensive devs, I feel obliged to reply. *This is my personal opinion, not an official project position*. Sorry it got a bit lengthy.

**TL;DR**
1. Generally no class projects - not sustainable.
2. Therefore no official documentation concerning class projects.
3. UToronto: Both sides should benefit. May continue. Some suggestions for improvement.

### We do not officially do class projects

We should not advertize Matplotlib as a class project, foremost because we don't have the capacity to handle large amounts of new contributors. - This includes not having formally written down guidelines, which would just bring more people to the idea.

If anybody wants to do this, we can discuss and decide on a case-by-case basis. (Contexts like experience, duration and scope will be different anyway.

### The case UToronto

We should only do this if both sides benefit.

#### Benefit for the matplotlib project

Quoting @tacaswell 
> from our (as the project) PoV the class is a source of relatively high-quality new contributors (we know they can program and they are externally motivated to do the work)

IMHO what we really need the most is not new contributors, but reviewers and core contributors. Many of the low hanging fruits are already picked or "boring" like code cleanup or documentation improvements.

The fact that students are externally motivated is actually negative. Someone with an intrinsic motivation is likely to stay longer and may become a core contributor.

Personally I doubt that there is a positive long-term effect for the project in terms of contributors (do we know if anybody from the project continued beyond the assignment?). Thus, only gain from this if the PRs are complex enough and high-quality so that reviewing these fixes and features takes less time than writing them ourselves.

#### How to proceed

Let's try to keep this going. Some suggestions to make it more smooth:

- @atafliovich should notify us contact us the project starts, so that we know what's coming.
- We as a project should decide how many groups we can reasonably support and communicate that back.
- We don't select or suggest specific suitable issues. This would be extra work, and anyway it's part of the student task to find these. - Whether we could tag issues better for new contributors is another, which is beyond this scope.
- Students should indicate that they are working in the scope of this assignment when they pick up an issue.
- We do not babysit the students, in particular even less than a regular new contributor; I expect the students to be reasonably competent developers, which is more than average. I.e. we do not have to give in-depth step-by-step descriptions what to do for an issue, probably not even links to the contribution docs are necessary, because the students should be prepared or work that out themselves. And they can ask *specifically* if they need some information.
- We do however aim at keeping the review process for these PRs going. Aspirationally,
  - We respond regularly to these PRs. Due to lack of reviewer resouces, unfortunately PRs fall through the cracks remain unresponded the reason for the increasing PR backlog.
  - We communicate early if we think the PR cannot be merged; e.g. either due to lack of review resouces, due to missing consensus on the way forward etc.

  This is better than what we currently achieve on average. But it does not help nor make for a good experience for the students if the don't get responses. OTOH this commitment is also the reason we should specify how many groups we can cope with.

- We do not need TA for early reviewing. Except for some extra attention, the PRs are not treated differently from any other PR.

- @atafliovich Students should be briefed how to interact with the project. Some things that I've seen
  - One group posted "Is anyone else currently working on this? If not I'll give it a try." one five issues. Two were cancelled later, two did not have any further action (apart from me telling them they can work on it). Additionally there was a discussion between groups who gets to work on an issue. I'd prefer, if the descission which group works on which issue is made offline. These messages trigger notifications and I check the issue just to see they are still in the process of deciding. Once a group has chosen an issue, they can announce it there.
  - There were two posts "Hey, I'm new to matplotlib and I am willing to contribute. Would this be a good first issue? If so what are some first steps in tackling this issue?" and "Hi, If no one is working on this issue than i would like to give it a shot. Can you please guide how can i address this issue? Thanks." Not sure if they were students, but if so I'd expect them to be able to start without the need of guidance (otherwise it's not the right issue for them).

**Comment by atafliovich** [comment]
Thank you for the very helpful feedback. I will comment on (3), as (1) and (2) are for the devs to discuss.

@timhoffm brings up several great points.

1) "The fact that students are externally motivated is actually negative. Someone with an intrinsic motivation is likely to stay longer and may become a core contributor."

I agree. Intrinsic motivation is certainly what you/we want. One of my goals of running these projects is getting the students exposed to and hopefully interested in the world of open source software. A survey in the beginning of this term showed that out of over 150 students, only 2 knew that contributing to open source is even a possibility. They are simply unaware of this community. Perhaps it is a generational issue...

2) " (do we know if anybody from the project continued beyond the assignment?)."

Excellent question. I don't know how to get this information other than surveying the students, say, a year after they graduate: the vast majority of the students in this course are in their final term and will graduate in June. Experience shows that such surveys have very low response rates. Tracking GitHub accounts is also not informative as most often students create separate GitHub accounts for school. Ideas welcome! :)

3) "Students should be briefed how to interact with the project."

This is the hardest one (for me) to address. The students already are more than just briefed on this. We talk about this in class. One of the requirements on the assignment is to follow contributors' guidelines for the project of their choice. The students are encouraged to run things by their TA before writing PRs, etc. The problem is... well, they are students. The authors of the examples you give (which are, of course, entirely inappropriate!) are likely students who neither come to class nor read/follow the assignment requirements. So, other than having every post be screened by the TAs, I don't know how to better address the issue.

**Comment by atafliovich** [comment]
Some more details about the course that may help in planning.

There are two project phases, following which the students can potentially submit PRs. 

1) In the middle of the term (next week) they submit a project deliverable in which they need to fix two bugs, no matter how small. This is where they learn the codebase and the process.

2) At the end of the term (April 3 this year) they submit a project deliverable in which they need to implement a new feature or fix a bug that is not trivial and requires a serious (well, serious for undergraduate students) amount of work.

Not every team will actually end up writing a PR, since it is not a requirement. We grade their "contributions" regardless of whether they are "real contributions". Having a PR merged gets them a non-negligible bonus, but is not a requirement to get a good grade. There is a higher chance that a small bug fix will reach you, as many teams will not get their final deliverables in shape (and the exams will be looming).

**Comment by story645** [comment]
>As I've probably started the topic with my comments

Nah, this has come up in the past too, which is why I opened the issue. If it really was just a one time thing, or honestly a one person thing), I wouldn't have bothered. 

>  This includes not having formally written down guidelines, which would just bring more people to the idea.

So I worry that if we don't write it down then instructors don't know that they should reach out to us & like I think the consensus is leaning towards a heads up is a good thing. (I'll put this on the agenda for Monday's call). Assigning open source work is becoming a thing in CS courses whether we say anything or not. 

>complex enough and high-quality so that reviewing these fixes and features takes less time than writing them ourselves.

This requires a ton of self confidence that a lot (probably most) folks don't have, especially most undergrads. Like I acknowledge that I'm not one of the folks reviewing all the things, but this reads like an insanely high bar (even if you don't mean it as such) and like there's so much research on how folks from underrepresented groups usually don't think they meet those bars even when they do, including CS students. I think testing the waters w/ a smaller investment like a bug fix is something we should be encouraging. 

 @atafliovich where does this course fall in their curriculum?

 >Not sure if they were students, but if so I'd expect them to be able to start without the need of guidance (otherwise it's not the right issue for them).

These are just as likely to be students interested in GSOC or run of the mill new contributors & queries like this should be directed to https://discourse.matplotlib.org/c/development/contribute/8

**Comment by timhoffm** [comment]
> So I worry that if we don't write it down then instructors don't know that they should reach out to us

In contrast, I worry that we'll get more contributions from that side, because people read it and realize we officially support classes. I would just wait till another class pops in and then deal with it.

> > complex enough and high-quality so that reviewing these fixes and features takes less time than writing them ourselves.
> This requires a ton of self confidence [...] this reads like an insanely high bar.

It's not something we have to write explicitly on the front page, but IMHO it should be more or less the criterion on which we support contributions from classes. You can lower the bar for individual cases, but it does not make sense for the project to invest time on babysitting lots of student.

> I'll put this on the agenda for Monday's call

Thanks. I really think this needs a discussion.

#### State of our review capacity
We've accumulated about 120 additional open PRs over the last year. Unfinished PRs are a source of frustration, both for authors as well as contributors. As long as we don't manage to process the current rate of new PRs, we should absolutely not do anything to increase the PR rate further.

For context, here is the number of reviews of PRs > 14000 (roughly last year) by the most common reviewers:

![grafik](https://user-images.githubusercontent.com/2836374/76164101-5930ca80-614c-11ea-887b-3493f3c05dac.png)

In the last year, I've spend more time than I should on Matplotlib for a long-term healthy balance between Matplotlib and work/sleep/relationship/other activities. Of that time, I used more than I wanted for reviews. I would like to provide more content myself. But it doesn't make sense to pile more and more PRs, so I spend most of my contribution time with reviewing. Now with our daughter, born, I will have less time for Matplotlib.

**To sum up:**
- our review capacity has not been sufficient in the past
- and it will probably be even lower in the future

I hope that this makes it more understandable, why I take a clear stand and set high expectations on PRs.

**Comment by anntzer** [comment]
> "everyone gets a little frustrated"

Just to be clear, some contribs are basically perfectly fine, e.g. #16694 tackled was well written, but needed a small change which I pointed out during review, the student understood it perfectly, the necessary changes were done, the thing was merged in just a couple of days.  I haven't carefully reviewed #16711 but at first sight it looks fine too, probably will need a round of review or two but nothing dramatic.  And these are PRs which fix real bugs.

On the other hand there are PRs which get stuck in limbo (that's not anything special about student PRs, I have PRs stuck in review for months and months), but then quite often even if we try to ressucitate them later the student is AWOL.  At least from my PoV, this is one of the most frustrating points: I don't really want to spend time reviewing a PR which will be slightly tricky, and for which there's quite some risks the other side just drops out after a while.

Really "bad" behavior is actually very rare (I can only remember a single instance of credit sniping), but leaves a very, very bad taste.

> Assigning open source work is becoming a thing in CS courses whether we say anything or not.

It would be silly to be so swamped by PRs that we have to restrict who can open PRs here, but I think this repo is not a playground; even if we're not paid to do this we're still trying to develop a professional plotting package.  (I'm also not volunteering for unlimited mentoring either -- GSOC notwithstanding.)

> there's so much research on how folks from underrepresented groups usually don't think they meet those bars even when they do

Even if I broadly agree with your point, please remember that not everyone here is North American, and may view this kind of statement rather differently.

**Comment by story645** [comment]
>Even if I broadly agree with your point, please remember that not everyone here is North American, and may view this kind of statement rather differently

Yeah even I struggled w/ how to frame it, and I think I can generalize it a bit here. I'm trying to get across my worry that setting high criteria on contributing might end up doing more harm than good in that there will still be plenty of folks who think their PRs meet that criteria even if the core contributors don't & the bar might scare off other potentially good contributors. 

Like as we're seeing, many of the frustrations in the review process are caused by contributors (students & not students) not reading the contributing guide/PR template/etc.

**Comment by atafliovich** [comment]
> @atafliovich where does this course fall in their curriculum?

@story645 This is a fourth-year course in a four-year program. The vast majority of these students will graduate within the next 4-8 months.

Regardless of what the project decides to do long term, is there a consensus on whether I should introduce some "filtering" this term?

**Comment by tacaswell** [comment]
We talked about this on today's dev call and I'm going to close this with no further action (this discussion will serve as as much documentation as we want to have).

I think everyone agrees that we don't want to, as a project, do anything special for classes in terms of prioritizing review / feedback, promising any sort of timeline or in adjusting our merge criteria for the students.  Any PR from students will be treated exactly the same as from any other (first time) contributor. 

We want to be a welcoming project that new contributors feel comfortable contributing to, while at the same time improve the code quality and professionalism of the library.  To that end we have a (sometimes slow) process for getting code (both bug fixes and new features into the library).  However our scarcest resources, as a project, is the time of our reviewers (as evidenced by our growing backlog of un-merged PRs).   We don't want to spend our time reviewing PRs that are going to be abandoned as soon as grades are in.  While @atafliovich is being considerate about interacting with us, I can imagine not all instructors would be and having Matplotlib serve as a distributed CS section would not be good outcome for anyone.

We would very much appreciate it if UToronto could provided more filtering on PRs this year to ones that you think have a good chance of getting in.  At a minimum, please make sure that the students read and fill out the templates :).  The class is already trying to teach the students these things (which at least I super appreciate), so (at least once) having a TA pre-approve groups PRs would be good.  For the final projects, please also only encourage groups that you think are self-motivated enough to see it through (even if it might take a while after the semester is over) to open PRs.

As @anntzer also points out, real bugs are being fixed by the students already this year and we have in the past gotten nice new features (3D Quiver comes to mind) from the UToronto class. I (with my personal hat on) hope that we can come to an arrangement where everyone continues to benefit!
